consensus/core/src/synchronizer.rs

```diff
@@ -992,17 +994,31 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
         missing_blocks: BTreeSet<BlockRef>,
         dag_state: Arc<RwLock<DagState>>,
     ) -> Vec<(BlocksGuard, Vec<Bytes>, AuthorityIndex)> {
-        const MAX_PEERS: usize = 3;
-
-        // Attempt to fetch only up to a max of blocks
+        // Preliminary truncation of missing blocks to fetch. Since each peer can have different
+        // number of missing blocks and the fetching is batched by peer, so keep more than max_blocks_per_sync
+        // per peer on average.
         let missing_blocks = missing_blocks
             .into_iter()
-            .take(MAX_PEERS * MAX_BLOCKS_PER_FETCH)
+            .take(2 * MAX_PERIODIC_SYNC_PEERS * context.parameters.max_blocks_per_sync)
             .collect::<Vec<_>>();
 
+        // Determine the number of peers to request missing blocks from.
+        let mut authorities = BTreeMap::<AuthorityIndex, Vec<BlockRef>>::new();
+        for block_ref in &missing_blocks {
+            authorities
+                .entry(block_ref.author)
+                .or_default()
+                .push(*block_ref);
+        }
+        let num_peers = authorities
+            .len()
+            .div_ceil((context.committee.size() - 1).div_ceil(MAX_PERIODIC_SYNC_PEERS));
```

Comment:
why is the `authorities.len().div_ceil((context.committee.size() - 1)` needed here?

Reply:
Added a comment. Basically we want to fetch against each peer for roughly the same number of authorities, not the same number of blocks.